lib/Frontend/CompilerInvocation.cpp

```diff
@@ -247,637 +246,758 @@ class ArgsToFrontendInputsConverter {
+  // perform (and dump the result of).
+  SmallVector<StringRef, 4> locations;
+  value.split(locations, ',');
+  
+  bool invalid = false;
+  for (auto location : locations) {
+    auto lineColumnStr = location.split(':');
+    unsigned line, column;
+    if (lineColumnStr.first.getAsInteger(10, line) ||
+        lineColumnStr.second.getAsInteger(10, column)) {
+      Diags.diagnose(SourceLoc(), diag::error_invalid_source_location_str,
+                     location);
+      invalid = true;
+      continue;
+    }
+    Opts.DumpScopeMapLocations.push_back({line, column});
+  }
+  
+  if (!invalid && Opts.DumpScopeMapLocations.empty())
+    Diags.diagnose(SourceLoc(), diag::error_no_source_location_scope_map);
+}
+
+FrontendOptions::ActionType
+FrontendArgsToOptionsConverter::determineWhatUserAskedFrontendToDo() const {
+  using namespace options;
+  const Arg *A = Args.getLastArg(OPT_modes_Group);
+  if (!A) {
+    // We don't have a mode, so determine a default.
+    if (Args.hasArg(OPT_emit_module, OPT_emit_module_path)) {
+      // We've been told to emit a module, but have no other mode indicators.
+      // As a result, put the frontend into EmitModuleOnly mode.
+      // (Setting up module output will be handled below.)
+      return FrontendOptions::EmitModuleOnly;
+    }
+    return Opts.RequestedAction; // no change
+  }
+  Option Opt = A->getOption();
+  if (Opt.matches(OPT_emit_object))
+    return FrontendOptions::EmitObject;
+  if (Opt.matches(OPT_emit_assembly))
+    return FrontendOptions::EmitAssembly;
+  if (Opt.matches(OPT_emit_ir))
+    return FrontendOptions::EmitIR;
+  if (Opt.matches(OPT_emit_bc))
+    return FrontendOptions::EmitBC;
+  if (Opt.matches(OPT_emit_sil))
+    return FrontendOptions::EmitSIL;
+  if (Opt.matches(OPT_emit_silgen))
+    return FrontendOptions::EmitSILGen;
+  if (Opt.matches(OPT_emit_sib))
+    return FrontendOptions::EmitSIB;
+  if (Opt.matches(OPT_emit_sibgen))
+    return FrontendOptions::EmitSIBGen;
+  if (Opt.matches(OPT_emit_pch))
+    return FrontendOptions::EmitPCH;
+  if (Opt.matches(OPT_emit_imported_modules))
+    return FrontendOptions::EmitImportedModules;
+  if (Opt.matches(OPT_parse))
+    return FrontendOptions::Parse;
+  if (Opt.matches(OPT_typecheck))
+    return FrontendOptions::Typecheck;
+  if (Opt.matches(OPT_dump_parse))
+    return FrontendOptions::DumpParse;
+  if (Opt.matches(OPT_dump_ast))
+    return FrontendOptions::DumpAST;
+  if (Opt.matches(OPT_emit_syntax))
+    return FrontendOptions::EmitSyntax;
+  if (Opt.matches(OPT_merge_modules))
+    return FrontendOptions::MergeModules;
+  if (Opt.matches(OPT_dump_scope_maps))
+    return FrontendOptions::DumpScopeMaps;
+  if (Opt.matches(OPT_dump_type_refinement_contexts))
+    return FrontendOptions::DumpTypeRefinementContexts;
+  if (Opt.matches(OPT_dump_interface_hash))
+    return FrontendOptions::DumpInterfaceHash;
+  if (Opt.matches(OPT_print_ast))
+    return FrontendOptions::PrintAST;
+  
+  // Symmetry violations:
```

Comment:
What does this mean?

Reply:
Good catch. I removed the comment; was useless.